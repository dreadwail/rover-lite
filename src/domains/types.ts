import { Action, AnyAction } from 'redux';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';

import { RoverSearchState } from '../state/types';

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

export type GenericThunkAction<T, S extends RoverSearchState = RoverSearchState> = ThunkAction<
  T,
  S,
  {},
  AnyAction
>;
export type PromiseThunkAction<
  T,
  S extends RoverSearchState = RoverSearchState
> = GenericThunkAction<Promise<T>, S>;

export type MapDispatchToProps<
  TDispatchProps,
  TOwnProps = {},
  TState extends RoverSearchState = RoverSearchState,
  A extends Action = Action
> = <E>(dispatch: ThunkDispatch<TState, E, A>, ownProps: TOwnProps) => TDispatchProps;
