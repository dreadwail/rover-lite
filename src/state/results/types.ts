export interface Badge {
  readonly content: string;
  readonly imageUrl: string;
  readonly title: string;
}

export interface Listing {
  readonly thumbnailUrl: string;
  readonly rank: number;
  readonly name: string;
  readonly badges: Badge[];
  readonly tagline: string;
  readonly neighborhood: string;
  readonly city: string;
  readonly state: string;
  readonly zip: string;
  readonly repeatClientCount: number;
  readonly reviewCount: number;
  readonly ratingAverage: number;
  readonly featuredReview: string;
  readonly currencySymbol: string;
  readonly price: number;
  readonly priceUnit: string;
  readonly webUrl: string;
}

export interface Results {
  readonly isFetching: boolean;
  readonly hasFailed: boolean;
  readonly isStale: boolean;
  readonly listings: Listing[];
}
