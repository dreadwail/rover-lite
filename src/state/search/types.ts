import { SearchType } from '../searchTypes/types';

export type SearchInputs = Record<string, any>;

export interface Search {
  readonly searchType: SearchType;
  readonly latitude: number;
  readonly longitude: number;
  readonly inputs: SearchInputs;
}
