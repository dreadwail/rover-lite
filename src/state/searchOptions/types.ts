export interface OptionCategory {
  readonly categoryId: string;
  readonly rank: number;
  readonly displayName: string;
}

export interface BasicOption {
  readonly type: string;
  readonly required: boolean;
  readonly label: string;
  readonly category: OptionCategory;
  readonly rank: number;
  readonly helpText?: string;
}

export interface DateOption extends BasicOption {
  readonly type: 'date';
}

export interface Choice {
  readonly value: any;
  readonly displayName: string;
}

export interface ChoiceOption extends BasicOption {
  readonly type: 'choice';
  readonly choices: Choice[];
}

export interface MultipleChoiceOption extends BasicOption {
  readonly type: 'multiple choice';
  readonly choices: Choice[];
}

export interface IntegerOption extends BasicOption {
  readonly type: 'integer';
  readonly minValue: number;
  readonly maxValue: number;
}

export interface BooleanOption extends BasicOption {
  readonly type: 'boolean';
}

export type Option =
  | BasicOption
  | DateOption
  | ChoiceOption
  | MultipleChoiceOption
  | IntegerOption
  | BooleanOption;

export const isBooleanOption = (option: Option): option is BooleanOption =>
  option.type === 'boolean';
export const isIntegerOption = (option: Option): option is IntegerOption =>
  option.type === 'integer';
export const isMultipleChoiceOption = (option: Option): option is MultipleChoiceOption =>
  option.type === 'multiple choice';
export const isChoiceOption = (option: Option): option is ChoiceOption => option.type === 'choice';
export const isDateOption = (option: Option): option is DateOption => option.type === 'date';
export const isBasicOption = (option: Option): option is BasicOption =>
  [isBooleanOption, isIntegerOption, isChoiceOption, isMultipleChoiceOption, isDateOption].every(
    check => !check(option)
  );

export type Options = Record<string, Option>;
export type OptionsByCategoryId = Record<string, Options>;

export interface SearchOptions {
  readonly isFetching: boolean;
  readonly hasFailed: boolean;
  readonly options: Options;
}
