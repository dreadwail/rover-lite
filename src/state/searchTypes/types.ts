export type SearchType =
  | 'overnight-boarding'
  | 'overnight-traveling'
  | 'drop-in'
  | 'doggy-day-care'
  | 'dog-walking';

export type Endpoints = { [T in SearchType]?: string };

export interface SearchTypes {
  readonly isFetching: boolean;
  readonly hasFailed: boolean;
  readonly endpoints: Endpoints;
}
