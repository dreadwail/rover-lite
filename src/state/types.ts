import { Results } from './results/types';
import { Search } from './search/types';
import { SearchOptions } from './searchOptions/types';
import { SearchTypes } from './searchTypes/types';

export interface RoverSearchState {
  readonly searchTypes: SearchTypes;
  readonly searchOptions: SearchOptions;
  readonly search: Search;
  readonly results: Results;
}
