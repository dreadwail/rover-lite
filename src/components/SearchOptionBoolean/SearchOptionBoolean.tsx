import React, { PureComponent } from 'react';

import { BooleanOption } from '../../state/searchOptions/types';

export interface SearchOptionBooleanProps {
  readonly optionKey: string;
  readonly option: BooleanOption;
  readonly value?: boolean;
  onChange(value: any): void;
}

export default class SearchOptionBoolean extends PureComponent<SearchOptionBooleanProps> {
  public render() {
    const { option, optionKey, value } = this.props;
    const { label } = option;
    return (
      <div className="form-group">
        <div className="form-check">
          <input
            className="form-check-input"
            type="checkbox"
            checked={!!value}
            onChange={this.handleChange}
            id={optionKey}
          />
          <label className="form-check-label" htmlFor={optionKey}>
            {label}
          </label>
        </div>
      </div>
    );
  }

  private readonly handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const { onChange } = this.props;
    onChange(event.target.checked);
  };
}
