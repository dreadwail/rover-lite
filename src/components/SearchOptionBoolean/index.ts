export * from './SearchOptionBoolean';
export { default } from './SearchOptionBoolean';
