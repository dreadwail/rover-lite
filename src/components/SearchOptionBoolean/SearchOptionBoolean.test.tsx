import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';

import { BooleanOption } from '../../state/searchOptions/types';

import SearchOptionBoolean, { SearchOptionBooleanProps } from './SearchOptionBoolean';

describe('SearchOptionBoolean', () => {
  let props: SearchOptionBooleanProps;
  let rendered: ShallowWrapper;

  beforeEach(() => {
    const option: BooleanOption = {
      type: 'boolean',
      required: false,
      label: 'test-label',
      category: {
        categoryId: 'test-category-id',
        rank: 0,
        displayName: 'Test Category',
      },
      rank: 0,
    };
    props = {
      optionKey: 'test-option-key',
      onChange: jest.fn().mockName('onChange'),
      option,
      value: true,
    };
    rendered = shallow(<SearchOptionBoolean {...props} />);
  });

  it('renders correctly', () => {
    expect(rendered).toMatchSnapshot();
  });

  describe('when the option is checked', () => {
    beforeEach(() => {
      rendered.find('input').simulate('change', { target: { checked: true } });
    });

    it('calls the onChange prop with the new value', () => {
      expect(props.onChange).toHaveBeenCalledWith(true);
    });
  });

  describe('when the option is unchecked', () => {
    beforeEach(() => {
      rendered.find('input').simulate('change', { target: { checked: false } });
    });

    it('calls the onChange prop with the new value', () => {
      expect(props.onChange).toHaveBeenCalledWith(false);
    });
  });
});
