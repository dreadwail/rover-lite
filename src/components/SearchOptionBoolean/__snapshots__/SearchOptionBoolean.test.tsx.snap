// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchOptionBoolean renders correctly 1`] = `
<div
  className="form-group"
>
  <div
    className="form-check"
  >
    <input
      checked={true}
      className="form-check-input"
      id="test-option-key"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      className="form-check-label"
      htmlFor="test-option-key"
    >
      test-label
    </label>
  </div>
</div>
`;
